src/main/java/graphql/execution/instrumentation/threadpools/ExecutorInstrumentation.java

```diff
@@ -0,0 +1,160 @@
+package graphql.execution.instrumentation.threadpools;
+
+import graphql.Assert;
+import graphql.TrivialDataFetcher;
+import graphql.execution.Async;
+import graphql.execution.instrumentation.SimpleInstrumentation;
+import graphql.execution.instrumentation.parameters.InstrumentationFieldFetchParameters;
+import graphql.schema.DataFetcher;
+import graphql.schema.DataFetchingEnvironment;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Executor;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import static graphql.execution.instrumentation.threadpools.ExecutorInstrumentation.Action.FETCHING;
+import static graphql.execution.instrumentation.threadpools.ExecutorInstrumentation.Action.PROCESSING;
+
+/**
+ * This instrumentation can be used to control on what thread calls to {@link DataFetcher}s happen on.
+ * <p>
+ * If your data fetching is inherently IO bound then you could use a IO oriented thread pool for your fetches and transfer control
+ * back to a CPU oriented thread pool and allow graphql-java code to run the post processing of results there.
+ * <p>
+ * An IO oriented thread pool is typically a multiple of {@link Runtime#availableProcessors()} while a CPU oriented thread pool
+ * is typically no more than {@link Runtime#availableProcessors()}.
+ * <p>
+ * The instrumentation will use the {@link graphql.execution.instrumentation.Instrumentation#instrumentDataFetcher(DataFetcher, InstrumentationFieldFetchParameters)}
+ * method to change your data fetchers so they are executed on a thread pool dedicated to fetching (if you provide one).
+ * <p>
+ * Once the data fetcher value is returns it will transfer control back to a processing thread pool (if you provide one).
+ * <p>
+ * This code uses {@link CompletableFuture#supplyAsync(Supplier, Executor)} and {@link CompletableFuture#thenApplyAsync(Function, Executor)} to transfer
+ * control between thread pools.
+ */
+public class ExecutorInstrumentation extends SimpleInstrumentation {
```

Comment:
can we mark this as Internal for now so that we can break it/remove if it doesn't work?

Reply:
done